Travis: enforce Python 3.4 support through linter

Sjors opened the pull request:
The minimum supported version of Python is 3.4 according to [dependencies.md](https://github.com/bitcoin/bitcoin/blob/master/doc/dependencies.md). This PR makes the Travis linter use this version in order to catch accidental use of modern syntax.

Sjors committed:
[test] Travis: enforce Python 3.4 support in functional tests

Make lint/check-doc.py Python 3.4 compatible.

Also add .python-version for pyenv which will cause tests with too
modern syntax to fail on developer machine rather than on Travis.

practicalswift commented:
Concept ACK

Good idea!

laanwj commented:
utACK

hebasto commented:
Concept ACK

Sjors commented:
* Travis build success: https://travis-ci.org/bitcoin/bitcoin/jobs/464372654
* Travis build (hopefully...) fails if I add commit 55474f5de3808bb235c620807e1ef83db1815354 with too modern syntax: https://travis-ci.org/Sjors/bitcoin/jobs/464366618

Sjors commented on .travis.yml:
```diff
@@ -104,6 +104,8 @@ jobs:
 # x86_64 Linux (no depends, only system libs, sanitizers: address/leak (ASan + LSan) + undefined (UBSan) + integer)
     - stage: test
+      language: python
+      python: '3.4' # Minimum supported version according to dependencies.md
```
In that case this probably won't work indeed... I thought we only used Docker on the Xenial host? I'm still waiting for Travis to build my test branch.

MarcoFalke replied:
https://github.com/bitcoin/bitcoin/blob/400a46c629a03ff5e01262c0301af3ad2c442a84/.travis.yml#L18 ;)

Sjors replied:
I see, trying a different approach...

MarcoFalke commented:
I mean of course strong Concept ACK if you figure out a way to install older versions (specific versions) of packages in ubuntu.

Sjors commented:
The only way to install older or newer versions of Python is through [pyenv](https://github.com/pyenv/pyenv#basic-github-checkout). I'd rather not add another dependency. 

Instead, a more practical solution is to bump the minimum Python version to 3.5, because our linters use that. It's available on Ubuntu Xenial, but we would lose support for (almost EOL) Trusty (Python 3.4).

In fact, `check-doc.py`, `optimize-pngs.py` , `gitian-build.py`, and `verify-commits.py` rely on Python 3.6 via `check_output(encoding='utf8')`. However only a small subset of developers need those scripts. I made trivial workaround for `check-doc.py` so Travis can use it with Python 3.5, but this workaround is not safe for the other scripts, see  #14128 

Alternatively we could modify the linter to work with Python 3.4 syntax. I personally can't be bothered, but I'll happily cherry-pick a solution :-)

I also added `.python-version` for `pyenv` which will cause tests with too modern syntax to fail on a developer machine before Travis. When using a script that does require Python >=3.6, you'll have to do `pyenv local 3.7.1`.

Sjors commented:
* Travis build fails if I add commit 5b742441bbc70a0b15a00d214321e79c58738395 with too modern syntax: https://travis-ci.org/bitcoin/bitcoin/jobs/464559620#L609

MarcoFalke commented:
The python 3.4 requirement is only for the tests, which might be run by users that compile from source. Personally I don't care about the python version of developer tools and linters and I don't mind not supporting 3.4 for those.

Sjors commented:
@MarcoFalke I initially thought the linter itself needed Python 3.5, but that's not true anymore. So in that case we don't have to bump the version, nice!

Travis build failure demo when you use too modern syntax: https://travis-ci.org/bitcoin/bitcoin/jobs/464581145#L609

Sjors commented on .travis.yml:
```diff
@@ -43,7 +43,7 @@ jobs:
       cache: false
       language: python
-      python: '3.6'
+      python: '3.4' # Oldest supported version according to doc/dependencies.md
```
The linter uses the available Python executable to lint the functional tests. It will throw an error if sees a syntax that it doesn't recognise, see above demo. So it's not necessary to run the functional tests themselves in Python 3.4, which would require adding pyenv as a dependency inside the Docker container and slow things down.

Sjors replied:
Also, though hopefully not needed, any developer who uses pyenv on their machine will detect any mistake that slips through the cracks once they run the functional test suite on master.

laanwj commented:
I *would* prefer having all tools work with the supported version of Python, for consistency for developers, but agree it's not urgent—having the tests pass is most important.

practicalswift commented:
@laanwj That's a good point. Agree.

Sjors commented:
I agree with that too, but it's non-trivial because #14128 had to introduce Python 3.6 syntax in those tools to deal with some sort of encoding hell.

laanwj commented:
> I agree with that too, but it's non-trivial because #14128 had to introduce Python 3.6 syntax in those tools to deal with some sort of encoding hell.

Yes, that particular functionality looks hard to replace. I didn't know that was Python 3.6 specific.

Sjors committed:
[test] functional framework: add CScript hex() for Python 3.4

test/functional/wallet_importmulti.py failed with:
AttributeError: 'CScript' object has no attribute 'hex'

Sjors commented:
Rebased and added a workaround for calling `.hex()` on a `CScript` array in #14886 (cc @jnewbery), which doesn't work before Python 3.5.

laanwj commented:
utACK 31926ee8cfc73501524dfa0fef2ccbaa786d6a00

jnewbery commented on .python-version:
```diff
@@ -0,0 +1 @@
+3.4.9
```
Does pyenv allow you to comment this file? It's not immediately obvious that this is for pyenv, how it's supposed to work, or when we should update this.

Sjors replied:
I'll look into that. At least you can find out via git-blame :-P

Sjors replied:
pyenv doesn't like it when you add a comment to `.python-version`, so I just updated the code style guidelines with an explanation in #15173.

jnewbery reviewed (commented):
utACK 31926ee8cfc73501524dfa0fef2ccbaa786d6a00. One request inline.

Thanks for catching my `hex()` mistake!

Sjors commented:
I missed a spot, see #15285
